cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricStoreRequestExecutor.java

```diff
@@ -111,9 +120,50 @@ private void computeProcessBusyness(CubeQuery query, TimeSeriesResponse.Builder
     }
   }
 
-  private Object computeQueueLength(CubeQuery query) {
-    // todo
-    return null;
+  private Object computeFlowletPending(CubeQuery query) throws Exception {
+    // Pending is processed by flowlet minus emitted into the queues it was processing from.
+
+    // trick: get all queues and streams it was processing from using group by
+    CubeQuery groupByQueueName =
+      new CubeQuery(new CubeQuery(query, "system.process.events.processed"),
+                    ImmutableList.of(Constants.Metrics.Tag.FLOWLET_QUEUE));
+    Map<String, Long> processedPerQueue = getTotalsWithSingleGroupByTag(groupByQueueName);
+
+    long processedTotal = 0;
+    long writtenTotal = 0;
+    for (Map.Entry<String, Long> entry : processedPerQueue.entrySet()) {
+      String name = entry.getKey();
+      // note: each has "input." prefix
+      QueueName queueName = QueueName.from(URI.create(name.substring("input.".length(), name.length())));
+      long written;
+
+      if (queueName.isQueue()) {
+        Map<String, String> sliceByTags = Maps.newHashMap(query.getSliceByTags());
+        // we want to aggregate written to the queue by all flowlets
+        sliceByTags.remove(Constants.Metrics.Tag.FLOWLET);
+        // we want to narrow down to specific queue we know our flowlet was consuming from
+        sliceByTags.put(Constants.Metrics.Tag.FLOWLET_QUEUE, queueName.getSimpleName());
+        written = getTotals(new CubeQuery(new CubeQuery(query, sliceByTags), "system.process.events.out"));
+
+      } else if (queueName.isStream()) {
+        Map<String, String> sliceByTags = Maps.newHashMap();
+        // we want to narrow down to specific stream we know our flowlet was consuming from
+        sliceByTags.put(Constants.Metrics.Tag.STREAM, queueName.getSimpleName());
+        // note: namespace + stream uniquely define the stream
+        // we know that flow can consume from stream of the same namespace only at this point
+        sliceByTags.put(Constants.Metrics.Tag.NAMESPACE, query.getSliceByTags().get(Constants.Metrics.Tag.NAMESPACE));
+        written = getTotals(new CubeQuery(new CubeQuery(query, sliceByTags), "system.collect.events"));
+      } else {
+        LOG.warn("Unknown queue type: " + name);
+        continue;
+      }
+      processedTotal += entry.getValue();
+      writtenTotal += written;
+    }
+
+    long pending = writtenTotal - processedTotal;
+
+    return new AggregateResponse(pending > 0 ? pending : 0);
```

Comment:
it was like that.. I guess it can - pending is not always correct :(. When you truncate stream - it can go bad :(